File: C_C++/q01493_fillingBox.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <string>
#include <utility>
#include <math.h>

using namespace std;
int N;
//vector<pair<int, int>> cube;
//vector<int> cube;
int cube[21];

void sort3int(int *a, int *b, int *c) {
  if (*a < *b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
  }
  if (*a < *c) {
    int tmp = *a;
    *a = *c;
    *c = tmp;
  }
  if (*b < *c) {
    int tmp = *b;
    *b = *c;
    *c = tmp;
  }
}

int devide(int l, int w, int h) {
  int result = 0;
  int min = l;

  
  if (l == 0 || w == 0 || h == 0){
    return 0;
  }
  if (w < min) min = w;
  if (h < min) min = h;

  int square = (int)(log(min) / log(2));
  int len = -1;

  for (int i = square; i >= 0; i--){
    if (cube[i] > 0) {
      result = 1;
      cube[i]--;
      len = pow(2, i);
      break;
    }
  }

  if (len == -1) {
    return -1;
  }
  
  int result1 = devide(l, w, h-len);
  int result2 = devide(l-len, w, len);
  int result3 = devide(len, w-len, len);

  if (result1 == -1 || result2 == -1 || result3 == -1) {
    return -1;
  }
  else {
    return 1 + result1 + result2 + result3;
  }
}

int main(void) {
  // Get Input
  int l, w, h;
  
  // vector<pair<int, int>> cube;

  cin >> l;
  cin >> w;
  cin >> h;

  cin >> N;

  int A, B;
  for (int i = 0; i < N; i++){
    cin >> A;
    cin >> B;
    cube[A] = B;
  }


  int result = devide(l, w, h);
  cout << result;  
}

File: python/q01493_fillingBox.py
import sys
from collections import deque
sys.setrecursionlimit(10**5)
input = sys.stdin.readline
cube = deque()

l, w, h = map(int, input().split(" "))
N = int(input())

for i in range(N) :
  A, B = map(int, input().split(" "))
  cube.appendleft([pow(2, A), B])

def devide(l, w, h, square) :
  # sort l, w, h
  h, w, l = sorted([l, w, h])
  
  while (l < square or w < square or h < square) and square > 1 :
    square /= 2

  if l == w == h :
    isPossible = False
    for cur in cube :
      if cur[1] == 0 :
        continue
      if cur[0] == l :
        cur[1] -= 1
        return 1
      elif cur[0] < l :
        isPossible = True
    if not isPossible :
      return -1
    square /= 2
  
  square = int(square)
  sum = 0
  result1 = devide(square, w, h, square)
  result2 = devide(l-square, w, h, square)

  if result1 == -1 or result2 == -1 :
    return -1
  else :
    sum = result1 + result2
    return sum
  
  # find largest cube that can be in this box
  for cur in cube :
    if cur[1] == 0 :
      continue
    if l == cur[0] and w == cur[0] and h == cur[0] :
      cur[1] -= 1
      return 1
    if h >= cur[0] :
      sum = 0

      result = devide(cur[0], w, h)
      if result == -1 :
        return -1
      else :
        sum += result

      result = devide(l-cur[0], w, h)
      if result == -1 :
        return -1
      else :
        sum += result

      return sum

  return -1

result = devide(l, w, h, 2**20)
print(result)